crypto/verification_sas_methods.go

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) 2020 Nikos Filippakis
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package crypto
+
+import (
+	"fmt"
+
+	"maunium.net/go/mautrix/crypto/olm"
+	"maunium.net/go/mautrix/id"
+)
+
+var allEmojis = [...]VerificationEmoji{
+	{'🐶', "Dog"},
+	{'🐱', "Cat"},
+	{'🦁', "Lion"},
+	{'🐎', "Horse"},
+	{'🦄', "Unicorn"},
+	{'🐷', "Pig"},
+	{'🐘', "Elephant"},
+	{'🐰', "Rabbit"},
+	{'🐼', "Panda"},
+	{'🐓', "Rooster"},
+	{'🐧', "Penguin"},
+	{'🐢', "Turtle"},
+	{'🐟', "Fish"},
+	{'🐙', "Octopus"},
+	{'🦋', "Butterfly"},
+	{'🌷', "Flower"},
+	{'🌳', "Tree"},
+	{'🌵', "Cactus"},
+	{'🍄', "Mushroom"},
+	{'🌏', "Globe"},
+	{'🌙', "Moon"},
+	{'☁', "Cloud"},
+	{'🔥', "Fire"},
+	{'🍌', "Banana"},
+	{'🍎', "Apple"},
+	{'🍓', "Strawberry"},
+	{'🌽', "Corn"},
+	{'🍕', "Pizza"},
+	{'🎂', "Cake"},
+	{'❤', "Heart"},
+	{'😀', "Smiley"},
+	{'🤖', "Robot"},
+	{'🎩', "Hat"},
+	{'👓', "Glasses"},
+	{'🔧', "Spanner"},
+	{'🎅', "Santa"},
+	{'👍', "Thumbs Up"},
+	{'☂', "Umbrella"},
+	{'⌛', "Hourglass"},
+	{'⏰', "Clock"},
+	{'🎁', "Gift"},
+	{'💡', "Light Bulb"},
+	{'📕', "Book"},
+	{'✏', "Pencil"},
+	{'📎', "Paperclip"},
+	{'✂', "Scissors"},
+	{'🔒', "Lock"},
+	{'🔑', "Key"},
+	{'🔨', "Hammer"},
+	{'☎', "Telephone"},
+	{'🏁', "Flag"},
+	{'🚂', "Train"},
+	{'🚲', "Bicycle"},
+	{'✈', "Aeroplane"},
+	{'🚀', "Rocket"},
+	{'🏆', "Trophy"},
+	{'⚽', "Ball"},
+	{'🎸', "Guitar"},
+	{'🎺', "Trumpet"},
+	{'🔔', "Bell"},
+	{'⚓', "Anchor"},
+	{'🎧', "Headphones"},
+	{'📁', "Folder"},
+	{'📌', "Pin"},
+}
+
+// VerificationEmoji describes an emoji that might be sent for verifying devices.
+type VerificationEmoji struct {
+	Emoji       rune
+	Description string
+}
+
+// GetSASVerificationNumbers generates the three numbers that need to match with the other device for a verification to be valid.
+func (mach *OlmMachine) GetSASVerificationNumbers(initUserID id.UserID, initDeviceID id.DeviceID,
+	acceptUserID id.UserID, acceptDeviceID id.DeviceID, transactionID string, sas *olm.SAS) ([3]uint, error) {
+	sasInfo := "MATRIX_KEY_VERIFICATION_SAS" +
```

Comment:
The `curve25519` key agreement method is apparently deprecated, the new one (`curve25519-hkdf-sha256`) is different: matrix-org/matrix-doc#2630 / https://matrix.org/docs/spec/client_server/unstable#hkdf-calculation

Reply:
We probably don't need to support the old method at all:

> New implementations are discouraged from implementing the `curve25519` method.